### PR TITLE
Federated tracing

### PR DESCRIPTION
**Pull request by tk26**
### :pencil: Description
This is to add Federated Tracing Instrumentation to `graphql-kotlin-federation`
Apollo's `federation-jvm` is taken as the reference (https://github.com/apollographql/federation-jvm/pull/25/)

### :link: Related Issues
Draft PR for https://github.com/ExpediaGroup/graphql-kotlin/issues/418

### CODE REVIEW

**Comment by tk26** [comment]
@dariuszkuc @smyrick Would love your feedback on this draft PR. As mentioned in the Description, this change refers implementation from `https://github.com/apollographql/federation-jvm` heavily (might as well be called it's Kotlin conversion)

If things look good, here's my TODO -
1. Give necessary credits to `federation-jvm`
2. Add tests

File: graphql-kotlin-federation/src/main/protobuf/reports.proto [hunk 1]
```diff
@@ -0,0 +1,493 @@
+// This file is copied over from the `apollo-engine-reporting` package, which, in turn, derives its information from
+// Apollo's cloud infrastructure. It is expected that as the protobuf definition changes within the underlying system
+// we are reporting to, these libraries will need to be updated in sync.
+// https://github.com/apollographql/apollo-server/blob/master/packages/apollo-engine-reporting-protobuf/src/reports.proto
```
Comment (smyrick) [inline graphql-kotlin-federation/src/main/protobuf/reports.proto]: Is there someway we can pull this file from NPM so we don't have to manage it here? https://www.npmjs.com/package/apollo-engine-reporting-protobuf

File: graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/tracing/FederatedTracingInstrumentation.kt [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+class FederatedTracingInstrumentation : SimpleInstrumentation() {
+    companion object {
+        const val HEADER_NAME: String = "apollo-federation-include-trace"
+        const val HEADER_VALUE: String = "ftv1"
+        const val EXTENSION_KEY: String = "ftv1"
+    }
+
+    override fun createState(parameters: InstrumentationCreateStateParameters): InstrumentationState? {
+        val context = parameters.executionInput.context
+        if (context is HttpRequestHeaders) {
```
Comment (dariuszkuc) [inline graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/tracing/FederatedTracingInstrumentation.kt]: This puts a requirement on the context object types. If we are going to go with this route we should create something very generic that has access to request and response objects instead.

**Comment by dariuszkuc** [comment]
Wondering whether we need all of that logic - haven't looked much into the actual code but technically once we create the underlying federated objects using federated type resolver then individual methods should be traceable by using default https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/tracing/TracingInstrumentation.java.

File: graphql-kotlin-federation/src/main/protobuf/reports.proto (hunk 1 above)
Comment (tk26) [reply graphql-kotlin-federation/src/main/protobuf/reports.proto]: Hmmm! Interesting. I can take a look, good thought 👍

**Comment by tk26** [comment]
> Wondering whether we need all of that logic - haven't looked much into the actual code but technically once we create the underlying federated objects using federated type resolver then individual methods should be traceable by using default [graphql-java/graphql-java:src/main/java/graphql/execution/instrumentation/tracing/TracingInstrumentation.java@`master`](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/tracing/TracingInstrumentation.java) [](/graphql-java/graphql-java/blob/HEAD@{2019-11-18T04:20:28Z}/src/main/java/graphql/execution/instrumentation/tracing/TracingInstrumentation.java).

Great point 👍 `TracingInstrumentation` seems to be able to do most of the job. We can make `FederatedTracingInstrumentation` extend `TracingInstrumentation` with a couple more changes - 
1. override `createState` to check if `apollo-federation-include-trace` header has `ftv1` value (Apollo Gateway's enforcement- https://www.apollographql.com/docs/apollo-server/federation/metrics/#how-tracing-data-is-exposed-from-a-federated-service)
2. Put the errors in `ftv1` extension (https://www.apollographql.com/docs/apollo-server/federation/metrics/#how-errors-work - paragraph 3)

The above 2 changes are necessary to make it work with Apollo Gateway - the question I have here is should we rather rename `FederatedTracingInstrumentation` as `ApolloFederatedTracingInstrumentation` for clarity?

**Comment by smyrick** [comment]
I think `FederatedTracingInstrumentation` since federation is created by apollo but has a spec that is not Apollo specific

**Comment by tk26** [comment]
> I think `FederatedTracingInstrumentation` since federation is created by apollo but has a spec that is not Apollo specific

The Apollo docsite references -
https://www.apollographql.com/docs/apollo-server/federation/metrics/#how-tracing-data-is-exposed-from-a-federated-service and https://www.apollographql.com/docs/apollo-server/federation/metrics/#how-errors-work

^ these indicate its very apollo-centric 🤔

**Comment by smyrick** [comment]
@tk26 I am going to close the draft until this is ready for review again. This is just to clean up our open PRs. Feel free to reopen when you need

**Pull request closed without merging**